lib/error/try_assign_test.cc

```diff
@@ -0,0 +1,30 @@
+#include "lib/error/try_assign.h"
+#include "gtest/gtest.h"
+#include "lib/error/status_or.h"
+
+using namespace error;
+
+const Status kDefault(Code(1), "default");
+
+TEST(TryAssignTest, CopyableStatusOr) {
+  constexpr auto fn = [](StatusOr<int> s) -> StatusOr<int> {
+    int a = 0;
+    TRY_ASSIGN(a, s);
```

Comment:
fwiw, I'm not a big fan of the TRY or TRY_ASSIGN. Both the name, and the use.

As a random reader who's never seen the code before.. try makes me think to catch / raise / exceptions. It is absolutely non-obvious what the code here is doing.

At the cost of more typing, I think that a plain if statement with return is certainly clearer. Especially if we can't make the macro safe in all cases / trivial to understand and change.

If you feel strongly about having the macro, I would recommend a name that makes obvious what they do.

Reply:
I feel very strongly about the macros themselves, but I'm happy to negotiate on the name. I'll give the rationale for both their existence and the names first:

First, why do we need these macros? At a high level, error handling is annoying and we've prohibited the language supported mechanism for dealing with errors (exceptions), so we should do everything we can to make error handling easy and concise. For control plane work I claim that the majority of code is error handling! I consider these macros to be an extension to the C++ language to bring it in line with the error handling in modern languages like rust and many functional languages.

You can argue that a single instance:

auto result = SomeFunctionReturningStatusOr();
const auto status = GetStatus(result);
if (IsError(status)) {
   return status;
}
const auto desired_value = GetValue(std::move(result));

...taken in isolation isn't terrible, however programming is mostly an exercise in not repeating oneself. Nearly any language feature that simplifies some common task comes at the expense of requiring reader knowledge. The question is always "is the knowledge cost worth paying?"

Unfortunately many functions contain more than one operation that can fail. Once you have 2+ copies of the above StatusOr<> handling code in a function, you start to have problems with variable naming. If the type of result is consistent, users are incentivized to reuse 'result' and 'status' which in turn incentivizes the removal of const qualifiers on variables. Both of these patterns (variable reuse and removal of const) are widely accepted to be bad code smells.

Assuming the type of 'result' varies or users attempt to use const judiciously, then we'll need multiple names for temporaries resulting in names like 'result2' and 'status2' (bad names are likely because the variables don't have strong meaning, long scope or obvious differences between the instances). Since the code is repetitive, copy-and-paste is likely which is particularly dangerous when combined with an incentive to use bad variable names (it is easy to check status2 when you meant status7).

You can argue that users can use anonymous blocks to reuse instances of 'result' and 'status', however this removes the possibility of creating making 'desired_value' const. In many cases the type of 'desired_value' is difficult to type (or impossible in the case of lambdas - this forces a conversion to std::function which sucks). At a minimum the combination of anonymous blocks with StatusOr<> results in forward declaration (always undesirable) of "desired_value' which the requires the user to come up with a safe default value to assign to 'desired_value'.

Finally, consumption of vertical space with low value content has significant negative impact on readability. I consider the minimum sequence that replaces TRY_ASSIGN() to be 5-6 lines long. Changing 5-6 repeated lines into a single line is a really big deal.

As for naming, I think a good argument can be made for RETURN_IF_ERROR and ASSIGN_OR_RETURN doing exactly what they say. This helps first time readers, a benefit that should not be underestimated. I chose TRY* because it mirrors the identical feature in rust and matches certain proposals for an equivalent "try 'expression';" syntax being proposed for C++2b to provide better support for explicit opt-in when calling exception throwing functions. The conciseness of TRY() is also attractive to me.

I've probably underestimated the surprise of TRY() since I've been thinking about the concept and looking at error handling in other languages for years now. Do you have any suggestions besides the obvious RETURN_IF_ERROR and ASSIGN_OR_RETURN?

Reply:
On my side, I was expecting your code snippet:
```
auto result = SomeFunctionReturningStatusOr();
const auto status = GetStatus(result);
if (IsError(status)) {
  return status;
}
const auto desired_value = GetValue(std::move(result));
```
To be more like:
```
auto result = SomeFunctionReturningStatusOr();
if (IsError(result)) {
  return result;
}
```
with uses looking more like:
```
printf("value is %d", GetValue(result));
```
eg, making the templated overloads nice to use directly?

That said, you have some good arguments, and you clearly feel strongly about them. I'm ok with having the macros, although still not convinced about the risk they bring in.
I really do like the conciseness of TRY, the names RETURN and ASSIGN however are significantly more readable. Sorry no strong advice here.

Reply:
Perhaps I was a little harsh in my example. Here's a slightly more precise version of your example.

auto result_or = SomeFunctionReturningStatusOr();
if (IsError(result)) {
  return GetStatus(result);
}

Moving GetValue(result_or) into a subsequent call is an improvement in many cases. I'd still argue that making users worry about variables of the form '*_or' and using GetValue() on every use is simply spreading around some of the complexity captured in the longer example.

I'm going to use ASSIGN_OR_RETURN and RETURN_IF_ERROR. I'm troubled that I never thought deeply about how descriptive these names are and instead just complained about the length.